lib/modules.nix

```diff
@@ -93,7 +93,10 @@ rec {
             res set._definedNames
         else
           res;
-      result = { inherit options config; };
+      result = {
+        inherit options;
+        config = removeAttrs config [ "_module" ];
```

Comment:
This is a huge breaking change and I don't understand why you would do this. `_module` is a perfectly valid part of `config` and being able to retrieve it is valuable, particularly when debugging.

Reply:
I think the naming (starting with `_`) was intended to convey that this is not a public interface so it's fine to remove/change  it.

Reply:
It's useful for debugging with `nix repl`.

Reply:
Without this change:
```
$ nix-instantiate --eval -A lib.maintainers.infinisil --strict --json | jq
{
  "_module": {
    "args": {},
    "check": true
  },
  "email": "contact@infinisil.com",
  "github": "infinisil",
  "githubId": 20525370,
  "keys": [
    {
      "_module": {
        "args": {
          "name": "[definition 1-entry 1]"
        },
        "check": true
      },
      "fingerprint": "6C2B 55D4 4E04 8266 6B7D  DA1A 422E 9EDA E015 7170",
      "longkeyid": "rsa4096/0x422E9EDAE0157170"
    }
  ],
  "name": "Silvan Mosberger"
}
```

I'd have to filter out all occurrences of `_module` manually otherwise. As far as I know `_module` is only used for the module system internally.

How do you use this for debugging?

Reply:
It lets me find accurately the values for the module arguments after invoking evalModules.
This is only using the module system for validation, right? Nixpkgs should use the raw data and not waste cycles on this validation. The raw data doesn't have `_module`. The validation can happen separately as part of the `lib` test suite.

Reply:
I guess that would work in this case (but missing fields won't have a default value anymore), but in general using the raw data doesn't work, as there could be some processing that changes it, adds defaults and such. This isn't the only place where one wouldn't have to filter `_modules` anymore:
```
$ rg '"_module"'
rename.nix
17:    (mkAliasOptionModule [ "environment" "checkConfigurationOptions" ] [ "_module" "check" ])

config/krb5/default.nix
40:      (attrName: attrValue: attrName != "_module" && attrValue != null)

services/monitoring/grafana.nix
74:  _filter = x: filterAttrs (k: v: k != "_module") x;

services/monitoring/prometheus/default.nix
54:  filterValidPrometheus = filterAttrsListRecursive (n: v: !(n == "_module" || v == null));

services/networking/nghttpx/default.nix
17:      (n: v: ("_module" != n) && (null != v) && (false != v))

services/monitoring/prometheus/exporters/mail.nix
14:          (filterAttrs (n: v: !(n == "_module" || v == null)) srv)
17:    ) (filterAttrs (n: _: !(n == "_module")) cfg.configuration)
```

Note that with this change it's still possible to get `config._module` from the `config` in the argument list for modules, which can be exposed through an explicit option if wanted.

Reply:
E.g. we could add `_moduleArgs = mkOption { default = config._module.args; internal = true; }` in a NixOS module

Reply:
Perhaps we can return it directly from `evalModules`? So we don't have `(evalModules {}).config._module` but `(evalModules {})._module`?
It still hides the args of submodules, but that can be solved by exposing it temporarily per case.
Seems acceptable that way. 
So only these things are left iiuc.
 - Could you make a separate issue or PR for that clean-up? It's not needed for the goal of this PR.
 - This PR needs to move the validation to the test suite. (Converting to JSON is not necessary now; nice to have)
 - Graham's concern about id correctness

Reply:
The checks are now done as part of `lib/tests/release.nix`